src: override v8 thread defaults using cli options

tomgco opened the pull request:
Based on the conversation in #4243 this implements a way to increase
and decrease the size of the thread pool used in v8. Which is defaulted
to `4`.

jasnell commented:
LGTM but definitely want to get @bnoordhuis' opinion /cc @nodejs/ctc

jasnell commented:
Oh, this would need to be added to the man page also.

tomgco commented:
> Oh, this would need to be added to the man page also.

Thanks for the heads up, I will update this PR with updates to documentation if more “LGTM”’s pour in.

Fishrock123 commented:
v8 doesn't have a flag for this?

tomgco commented:
> v8 doesn't have a flag for this?

In [`node.cc`](https://github.com/nodejs/node/pull/4344/files#diff-cd53544f44aab2c697bcd7b6a57f23ccR4218) we explicitly create a default platform in V8 with a hard-coded value of `4`, If set to `0` then v8 would choose an appropriate size of the thread pool based on the number of online processors, in #4243 it was suggested to set a length via CLI options where a suitable default could be decided later on, but at the same time allow to decide based on their own workloads.

But this is an implementation detail, which only embedder’s can change AFAIK.

bnoordhuis commented:
I checked V8's `DefaultPlatform` implementation and it actually restricts the thread pool to kMaxThreadPoolSize (4) threads.  So with this PR you could lower the number of threads but you can't increase it.

bnoordhuis commented on src/node.cc:
```diff
@@ -3398,6 +3402,8 @@ static void ParseArgs(int* argc,
       new_v8_argv[new_v8_argc] = "--help";
       new_v8_argc += 1;
+    } else if (strncmp(arg, "--v8-pool-size=", 15) == 0) {
+      v8_thread_pool_size = arg + 15;
```
Why not parse the string immediately?

tomgco replied:
I haven’t written much c++ before, but the rationale behind this is that if we parsed the string immediately we would have to set the value of the `int` as it is not a pointer it would have to be `-1` which I am not sure makes sense?

If it does then I can change it to be `-1` if not found and then parse the string immediately.

cjihrig replied:
Couldn't you set the pool size to the default of 4 when it's declared and then update the value here? I **think** that is what @bnoordhuis is getting at.

jasnell replied:
@tomgco ... essentially, rather than defining `v8_thread_pool_size` as a `const char*`, you can create it as an int with a default value of `v8_default_thread_pool_size`, then, in the check for `--v8-pool-size`, parse the string immediately, if it's within the appropriate int range ( > 0, <= the current maximum ... which is 4) set `v8_thread_pool_size` to the new value. You can then avoid the `pool_size` declaration later and simply `default_platform = v8::platform::CreateDefaultPlatform(v8_thread_pool_size)`.

tomgco replied:
@jasnell Thanks for the hand, I am not going to include the checks that the `int`'s are in an appropriate range as V8 already handles this internally (https://github.com/v8/v8/blob/master/src/libplatform/default-platform.cc#L68-L72), It will save on maintenance time if these values or constants change, do you think that this is a wise thing to do?

tomgco commented:
> I checked V8's DefaultPlatform implementation and it actually restricts the thread pool to kMaxThreadPoolSize (4) threads. So with this PR you could lower the number of threads but you can't increase it.

Thanks, I will make sure to include that in the documentation.

indutny commented:
Needs docs, otherwise LGTM

Fishrock123 commented on src/node.cc:
```diff
@@ -4202,8 +4208,14 @@ int Start(int argc, char** argv) {
+    pool_size = atoi(v8_thread_pool_size);
+  }
+
+  default_platform = v8::platform::CreateDefaultPlatform(pool_size);
```
Should we maybe emit warnings here then if the size is too big/small?

i.e. that the threadpool max is 4 if it is larger than that, and also something else if the user specified `<1`

jasnell replied:
I would simply say that if the value is not within an acceptable range, the nearest acceptable value is selected... essentially, anything less than 0 is treated as 0, anything greater than the current max is treated as the max. If we add that to the documentation, no warning should be necessary.

tomgco replied:
> I would simply say that if the value is not within an acceptable range, the nearest acceptable value is selected... essentially, anything less than 0 is treated as 0, anything greater than the current max is treated as the max. If we add that to the documentation, no warning should be necessary.

This is already handled in `default-platform.cc` in V8 https://github.com/v8/v8/blob/master/src/libplatform/default-platform.cc#L68-L72, Any values less than 1 is set to the number of online processors, anything larger than the current kMaxThreadPoolSize is limited to it.

ofrobots commented:
> Thanks, I will make sure to include that in the documentation.

The commit log message also needs to be updated. It would be better not to include specific numbers in the docs, as they may change if V8 changes `DefaultPlatform::kMaxThreadPoolSize` in the future.

tomgco commented:
> The commit log message also needs to be updated. It would be better not to include specific numbers in the docs, as they may change if V8 changes DefaultPlatform::kMaxThreadPoolSize in the future.

Sorry for the delay in the reply for this, holidays and all, will start fixing all of the issues pointed out now.

tomgco commented:
@ofrobots @jasnell @Fishrock123 I have authored the commit with a new message, updated the `--help`  output to remove any hard coded values and updated the man page.

ofrobots commented:
> I checked V8's DefaultPlatform implementation and it actually restricts the thread pool to kMaxThreadPoolSize (4) threads. So with this PR you could lower the number of threads but you can't increase it.

FYI, I am trying to fix this limitation upstream in this CL: https://codereview.chromium.org/1553653002/

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/1244/

jasnell commented:
@tomgco ... ping.. have you been able to look at the most recently comments...

tomgco commented:
> @tomgco ... ping.. have you been able to look at the most recently comments...

Sorry about the delay, I have been without Internet. I am looking at them now. Thanks for the reminder.

tomgco commented:
@ofrobots @jasnell @Fishrock123 I have addressed all comments and updated the commit as well as the wording of the commit message.

cjihrig commented on src/node.cc:
```diff
@@ -168,6 +169,7 @@ static uv_async_t dispatch_debug_messages_async;
 static node::atomic<Isolate*> node_isolate;
 static v8::Platform* default_platform;
 
+int v8_thread_pool_size = v8_default_thread_pool_size;
```
Would it work to ditch `v8_default_thread_pool_size`, and just initialize this to `DefaultPlatform::kMaxThreadPoolSize`, like @ofrobots suggested?

I think you'll want to group this variable with the variables declared starting at line 134.

tomgco replied:
> Would it work to ditch v8_default_thread_pool_size, and just initialize this to DefaultPlatform::kMaxThreadPoolSize, like @ofrobots suggested?

That would depend if we wanted to change our defaults based on upstream v8 changes, or have a default value that is controlled from within core. Let's say v8 changed kMaxThreadPoolSize to MAX_INT, I am sure that we would not want to set the default based on this?

> I think you'll want to group this variable with the variables declared starting at line 134.

Thanks! Was looking for a nicer place. Will update that now.

cjihrig replied:
Good point. Leaving it as is is fine.

tomgco commented:
@cjihrig Ok, pushed the grouping of the variables.

cjihrig commented on src/node.cc:
```diff
@@ -144,6 +144,8 @@ static const char** preload_modules = nullptr;
 static bool debug_wait_connect = false;
 static int debug_port = 5858;
+static const int v8_default_thread_pool_size = 4;
+int v8_thread_pool_size = v8_default_thread_pool_size;
```
Seems like this should probably be `static` for consistency.

cjihrig commented:
LGTM. It would be nice if there were a way to test this though.

jasnell commented:
LGTM, can't think of a reliable way of testing tho

tomgco commented:
> LGTM, can't think of a reliable way of testing tho

I have been thinking about this. The only way that I have thought of so far is (If V8 allows it) to expose the thread pool size in `process`. And then assert against that, but to me this doesn't seem a nice option. I am welcome to suggestions though.

jasnell commented:
That's likely not ideal. I'd say let's go ahead and get this landed. If we can come up with a way of testing it later, then great.

jasnell commented:
@tomgco ... looks like this one should have landed but fell through the cracks. Can you rebase and update and I'll get a CI going.

tomgco commented:
@jasnell No problem, I have just rebased this on master and should be ready to kick off the CI again.

romankl commented:
And here's a CI run: https://ci.nodejs.org/job/node-test-pull-request/2019/

jasnell commented:
Landed in 03e07d3

Fishrock123 commented:
I'm a little concerned that the flag is not descriptive enough and should be renamed to `--v8-thread-pool-size` before we have a release with it. Thoughts?

jasnell commented:
sounds ok to me. I'm good either way tho.
